design/230728_tinkerbell_workflow_execution.md

```diff
@@ -0,0 +1,491 @@
+
+## Goals/Non-goals
+
+**Goals**
+* Simplify the Tink Agent - Tink Server communication to what's necessary.
+* Define the components involved in Workflow execution and their responsibilities.
+* Provide clear behavioral expectations in failure cases.
+
+**Non-goals**
+* Introduce new functionality beyond that specified in the v1alpha2 API unless necessary for the new Workflow execution proposal.
+
+## Existing Solution
+
+Tinkerbell defines 3 core components involved with Workflow execution: Tink Worker, Tink Server and Tink Controller. Tink Worker runs on the target device and executes Workflow actions reporting the result of actions to Tink Server. On startup, Tink Worker connects to Tink Server providing a unique identifier that Tink Server uses to retrieve Workflows. As Tink Server receives the result of executing actions from Tink Worker, it updates the Workflow resource.
+
+Tink Controller is responsible for preparing Workflows resources using Templates. When a user submits Hardware, Workflow and Template resources to the system, Tink Controller renders the Template storing the output on the Workflow so it can be updated as Workflow actions execute. Tink Controller bears no other responsibility.
+
+Tink Worker communicates with Tink Server over gRPC. There are a handful of obsolete RPCs that are unimplemented in Tink Server since the transition to the Kubernetes backend (see Appendix A for existing protobuf).
+
+## Proposal
+
+### Summary
+
+The components used in Workflow execution will remain similar to the existing solution. Tink Agent (renamed from Tink Worker) will run on the device to be provisioned and is responsible for executing Workflows. Tink Server will instruct Tink Agent to run and cancel Workflows, and will update Workflow status based on Workflow execution Events published by Tink Agent. Tink Controller will prepare Workflows for execution and monitor them for erroneous states taking any necessary remedial action. The architecture draws similarities with the [Scheduler-Agent-Supervisor](https://learn.microsoft.com/en-us/azure/architecture/patterns/scheduler-agent-supervisor) pattern.
+
+### Workflow States
+
+<p align="center">
+  <img alt="[State Transitions]" src="./images/230728_tinkerbell_workflow_execution/state.png" />
+</p>
+
+#### Why do we need the Scheduled state?
+
+The Scheduled state is an intermediary state that indicates a Workflow has been dispatched to Tink Agent but hasn't started. Tink Controller will monitor Workflows in a Scheduled state and transition them to Failed if a timeout trips.
+
+Workflows may also be rejected by Tink Agent explicitly for cases such as a double dispatch. In the event a Workflow is rejected it will transition back to Pending. Workflow rejection is discussed later.
```

Comment:
Shouldn't actually this be done by the tink-controller? since it aware of the state of the tink-agent anyway. This would lead to rejecting the dispatched workflow quite early in the chain

Reply:
I think the workflow rejection discussion might clarify some things. Tink Agent doesn't do anything other than publish a rejection event. Tink Server handles that event by transitioning the Workflow.

I don't think Tink Controller needs to be involved with explicit handling of events, that's a Tink Server thing in the design currently.